pkg/bucket/lifecycle/filter.go

```diff
@@ -18,76 +18,73 @@ package lifecycle
 
 import (
 	"encoding/xml"
+	"strings"
 
 	"github.com/minio/minio-go/v6/pkg/tags"
 )
 
-var (
-	errInvalidFilter = Errorf("Filter must have exactly one of Prefix, Tag, or And specified")
-)
+var errInvalidFilter = Errorf("Filter must have exactly one of Prefix, Tag, or And specified")
+
+func isAllowed(prefix string, and *And, tag *tags.Tag) bool {
+	return (prefix != "") != (and != nil) != (tag != nil) || // prefix XOR and XOR tag.
+		(prefix == "" && and == nil && tag == nil) // empty prefix allowed.
+}
 
 // Filter - a filter for a lifecycle configuration Rule.
 type Filter struct {
-	XMLName xml.Name `xml:"Filter"`
-	Prefix  string
-	And     And
-	Tag     tags.Tag
+	XMLName xml.Name  `xml:"Filter"`
+	Prefix  string    `xml:"Prefix"`
+	And     *And      `xml:"And,omitempty"`
+	Tag     *tags.Tag `xml:"Tag,omitempty"`
 }
 
-// MarshalXML - produces the xml representation of the Filter struct
-// only one of Prefix, And and Tag should be present in the output.
-func (f Filter) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
-	if err := e.EncodeToken(start); err != nil {
-		return err
+func (f Filter) getPrefix() string {
+	if f.Prefix != "" {
+		return f.Prefix
 	}
 
-	switch {
-	case !f.And.isEmpty():
```

Comment:
Actually @vadmeste went to great lengths to ensure that we do this without using pointers, as for `Tag *tags.Tag` it makes sense since its a larger composite data structure. 

But not for And so we should keep it.